File: week2/python/fibonacci_sum_last_digit.py
# Uses python3

def fibonacci_sum_optimized(n):
    if n <= 1:
        return n

    new_n = (n + 2)%60
    if new_n == 1:
        return 0
    elif new_n == 0:
        return 9

    previous = 0
    current  = 1

    for _ in range(2, new_n + 1):
        c = current + previous
        c = c%10
        current, previous = c, current
    
    if c != 0:
        return c-1
    else:
        return 9

if __name__ == '__main__':
    n = int(input())
    print(fibonacci_sum_optimized(n))

File: week2/C++/fibonacci_sum_last_digit.cpp
#include <iostream>

int fibonacci_sum_optimized(long long n) {
    if (n <= 1)
        return n;
    int new_n = (n+2)%60;
    if (new_n == 1)
        return 0;
    else if (new_n == 0)
        return 9;


    long long previous = 0;
    long long current  = 1;
    long long c;
    for (long long i = 2; i < new_n + 1; ++i) {
        c = current + previous;
        c = c % 10;
        previous = current;
        current = c;
    }
    if (c != 0)
        return (c-1);
    else
        return 9;
}

int main() {
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum_optimized(n) << "\n";
}
